pkg/container-disk/container-disk.go

```diff
@@ -230,6 +230,18 @@ func GenerateKernelBootInitContainer(vmi *v1.VirtualMachineInstance, imageIDs ma
 	return generateKernelBootContainerHelper(vmi, imageIDs, podVolumeName, binVolumeName, true)
 }
 
+func GenerateTpmDirInitContainer(vmi *v1.VirtualMachineInstance, image string) *kubev1.Container {
+	return &kubev1.Container{
+		Name:    "tpm-dir",
+		Image:   image,
+		Command: []string{"mkdir", "-p", "/var/run/kubevirt-private/libvirt/qemu/swtpm"},
```

Comment:
Hm, init containers are in principle a little bit problematic and have for instance quite a bit of influence on the vm startup time.

Can virt-launcher just create the directory or virt-handler?

Reply:
The problem with such old code is that I don't quite remember why I did things that way!
I do remember there was a good reason for this, might be SELinux-related, I wish I had been more comment-generous...
I'll test without it and keep you posted!

Reply:
@jean-edouard can you please explain why you create this directory?

Reply:
This is where the swtpm state file will be created. We want to mount the persistent storage there so the state will be automatically preserved, without having to resort to symlinks that SELinux won't like.

Reply:
Sorry, it is still unclear to me. Isn't the state stored in the PVC? Isn't the mount and SELinux handled directly by k8s

Reply:
`compute` will request to have PVC volume mounted under `/var/run/kubevirt-private/libvirt/qemu/swtpm`. I believe the directory needs to exist for the container to start successfully. 
Actually it's possible that the directory does get created by k8s if absent, but with the wrong permissions, in which case this might actually be something we can fix in virt-handler, as suggested by Roman. I will test that.

Reply:
Yeah it's a permission thing... I just tested without the init container and got `panic: open /var/run/kubevirt-private/libvirt/qemu.conf: permission denied`
Is it a CRI bug? Does it really make sense to implement the logic for virt-handler to intervene before virt-launcher touches anything under `/var/run/kubevirt-private/libvirt/`?

Reply:
Maybe adding `fsGroup: 107` to virt-launcher pod can solve the isue?

Reply:
mh.. can you identify where the permissions are wrong? If it is on the directory or on the files? Do we set the fsGroup in virt-launcher, [here](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods) you can see in more details how the permission work

Reply:
Ok so, without the init container, `/var/run/kubevirt-private/libvirt` is owned by `root` (bad) and has the SELinux label `container_file_t` (good). I believe `fsGroup` (which we don't set) would probably fix that (but I don't know if/how things like PSA restrict `fsGroup`).
However, `/var/run/kubevirt-private/libvirt/qemu` is not only owned by root, but also has the SELinux label `system_u:object_r:nfs_t:s0`, which is a bigger problem (I use the nfs-csi storage class for testing).

Reply:
IIRC, nfs-csi isn't properly setting the ownership. @akalenyu do you recall? It might be a problem of that particular storage provider

Reply:
So no-provisioner NFS will not respect fsGroup at all, but nfs-csi should https://github.com/kubevirt/kubevirtci/blob/main/cluster-provision/k8s/1.26-centos9/manifests/nfs-csi/csi-nfs-driverinfo.yaml#L11

Reply:
After setting fsGroup to 107, both directories are now owned by `root:qemu`, but `libvirt` has a mode of `drwxr-sr-x`, which doesn't allow qemu to create files under them!
`fsGroup` also doesn't fix the SELinux label issue (unsurprisingly).

Reply:
Can you try to set the PVC explicitly not read-only? The SELinux seems to be a problem of the storage driver :/. You could give a try with ceph, you can set `export KUBEVIRT_STORAGE=rook-ceph-default`  before `cluster-up`

Reply:
I need a storage class that supports FS + RWX. I'll see if I can make cephfs work.

Reply:
With fsGroup=107 and rook-cephfs as a storage class:
```
/var/run/kubevirt-private/libvirt:
total 0
drwxr-sr-x. 3 root qemu system_u:object_r:container_file_t:s0:c547,c660 19 Feb 15 15:06 qemu

/var/run/kubevirt-private/libvirt/qemu:
total 0
drwxrwsr-x. 2 root qemu system_u:object_r:container_file_t:s0:c547,c660 0 Feb 15 15:06 swtpm
```
The `libvirt` directory is still a problem:
```
bash-5.1$ touch /var/run/kubevirt-private/libvirt/test
touch: cannot touch '/var/run/kubevirt-private/libvirt/test': Permission denied
```

Reply:
mh and can you add also the `/var/run/kubevirt-private/libvirt` under the VolumeMount and then the other subdirectories? It's just a guess, not sure if it's possible

Reply:
> mh and can you add also the `/var/run/kubevirt-private/libvirt` under the VolumeMount and then the other subdirectories? It's just a guess, not sure if it's possible

Do you mean adding all the intermediate directories as emptydirs? It might work, but that's not pretty!

Reply:
Well, it is a way to say that you want also the upper directory writable. Otherwise, k8s/CSI might assume you don't need to write into it and set it readonly

Reply:
Testing now. If it works, it sounds like you prefer that approach over the init container trick. @rmohr WDYT?

Reply:
> Testing now. If it works, it sounds like you prefer that approach over the init container trick. @rmohr WDYT?

Yes, I'd definitely prefer this to the init container

Reply:
Seems to work fine with cephfs, will test nfs after. However, it does require `fsGroup` to be set, as the swtpm directory doesn't allow writes from _other_.
@xpivarc @acardace from a PSA point of view, is it acceptable to set `fsGroup` to 107 under the pod security context? Thanks

Reply:
PSA should not be a problem here. But I remember we did not want to use fsGroup in the past. I can't remember why.

Reply:
I think we were setting selinux contexts and permissions before from virt- handler. It is simple and reliable and works for all storage classes. Unless there are some side effects which I am not aware of which we want to avoid nowadays I would recommend virt-handler as the place to ensure that permissions are correct.

Reply:
@rmohr do you have an example of that? I tried grepping around for it and couldn't find anything.
Also, implementing a hook in the virt-launcher code to wait for virt-handler to come and fix things doesn't really feel "cheaper" to me that a couple of emptyDirs, or even a small init container...

Reply:
@jean-edouard did it work also for ceph-csi?
If you don't like listing too many directories with emptyDir, do you think it would be possible to create only the upper directory and then let virt-launcher create the subdirectories at start time?

Reply:
@jean-edouard maybe I missed the part that this directory needs to be ready when libvirt starts already, and not when qemu starts? Is that the case?

Reply:
> @jean-edouard maybe I missed the part that this directory needs to be ready when libvirt starts already, and not when qemu starts? Is that the case?

The directory needs to be ready even before libvirt starts, as virt-launcher first prepares qemu.conf in there.

Reply:
> @jean-edouard did it work also for nfs-csi? If you don't like listing too many directories with emptyDir, do you think it would be possible to create only the upper directory and then let virt-launcher create the subdirectories at start time?

Yes it did work for nfs-csi. Every sub-directory that's not explicitly created via init-container/emptyDir will not be writable. In this case I'm pretty sure we use every sub-directory of `/var/run/kubevirt-private/libvirt/qemu/swtpm`.
However, I didn't have to do anything for `/var/run/kubevirt-private/var/lib/swtpm-localca` since nothing writes to `var/` or `lib/`.

Reply:
>> @jean-edouard maybe I missed the part that this directory needs to be ready when libvirt starts already, and not when qemu starts? Is that the case?

> The directory needs to be ready even before libvirt starts, as virt-launcher first prepares qemu.conf in there.

If that is the case, then I am happy with any solution which does not use an additional init container. So sgtm. Thanks. :+1: